### ISSUE TITLE
Validate path parameters?

### ISSUE THREAD

**Issue by joshuajabbour** [BigstickCarpet/swagger-express-middleware#5]
Is there validation of parameter data? Right now, if I have this spec:

```yaml
  /pets/{id}:
    parameters:
      - name: id
        in: path
        description: Resource ID
        required: true
        type: integer
        format: int32
    get:
      description: Finds a pet by ID
      operationId: findPetById
      responses:
        default:
          description: Returns the pet data
          schema:
            $ref: pet
```

when using an invalid URL I get: `swagger:middleware GET /pets/abc matches Swagger path /pets/{id}` which is definitely not true.

Is this just something that hasn't been implemented yet? Or am I missing it?

**Comment by joshuajabbour** [comment]
Yeah, diving into this a bit more, there definitely seems to be something up with path parameters. They never hit `parsePathParam`, and `req.params` never gets filled with the data. I'll keep digging tomorrow, but I'm not sure where the issue could be...

**Comment by BigstickCarpet** [comment]
Are you passing your Express App to the middleware?  That's the only way that the `parsePathParam` method will ever get called.  This is because Express uses special "route param middleware" to parse route parameters.  Details are [explained here](https://github.com/BigstickCarpet/swagger-express-middleware/blob/master/docs/middleware/parseRequest.md#phase-3---path-parsing).

**Comment by BigstickCarpet** [comment]
Also, to clarify the `swagger:middleware GET /pets/abc matches Swagger path /pets/{id}` message that you are seeing.   That is prior to parameter validation.  It's just letting you know that it figured out which path in your Swagger spec corresponds to the URL of the request.   Later on, you'll see a different message like `Parsing the "id" path parameter`.  That's when it does the parameter validation and will throw an error if it's not an integer.

**Comment by joshuajabbour** [comment]
Yeah, I'm def passing my app into the middleware (called `swagger` here):

```
swagger(spec, app, function(err, middleware, api, metadata) {
```

I'll do a little more debugging, but I basically started with the sample2 structure. And my breakpoints in the `parsePathParam` never get hit.

**Comment by joshuajabbour** [comment]
Here are the debug messages I get:
```
  swagger:middleware GET /pets/abc matches Swagger path /pets/{id} +39s
  swagger:middleware Parsing 1 request parameters... +1ms
  swagger:middleware Validating Accept header (application/json) +0ms
  swagger:middleware Validating Content-Length header (NaN) +0ms
```

Yet the route never errors out, and `req.params` never gets filled.

**Comment by BigstickCarpet** [comment]
Can you post your Swagger spec and your JavaScript file?   That way I can reproduce the problem on my end and figure out what's up.  If it's a bug in my code, I'll get it fixed ASAP.

**Comment by joshuajabbour** [comment]
Here's a basic test: https://gist.github.com/joshuajabbour/9beeba181d309c7ada58

What am I doing wrong there?

**Comment by BigstickCarpet** [comment]
Wow!  Didn't expect you to post a complete example including a unit test.  Thanks!

There are two reasons this code isn't working:
1) The Express App isn't being passed to the middleware.   This fixes that problem:

````javascript
var app = express();
swagger(spec, app, function(err, middleware) {
````

2) No path params are being passed to `app.use()`.  Express needs to know that there's a path param that needs to be parsed.  This fixes that problem:

````javascript
app.use('/pets/:id', function (req, res, next) {
  // debugger;
  res.json(req.params);
});
````

With those two changes in place, the unit test passes.  I'll admit that this isn't exactly straightforward at all.  I _really_ wish Express didn't require us to jump through so many hoops to parse path parameters.

On a related note... Express re-creates the `req.params` object for _each_ middleware in the pipeline.  So even if a param is parsed for one middleware, it _won't_ be parsed for the next one in the pipeline unless a path param was specified when that middleware was registered.  Here's an example of that:

````javascript
// The "id" param will be parsed for this middleware 
app.use('/pets/:id', function (req, res, next) {
  console.log(req.params.id);   // => outputs a number
  next();
});

// The "id" param will NOT be parsed for this middleware,
// even though it WAS parsed for the previous middleware in the pipeline
app.use(function(req, res, next) {
  console.log(req.params.id);   // => outputs a string
  next();
});
````

**Comment by joshuajabbour** [comment]
Well, I mostly wrote the test so that I could see if I could narrow it down before bugging you. And to strip away a lot of stuff I'm doing. But in my actual server code, I *am* passing the `app` into the middleware. I just didn't port that over here, stupidly.

The second issue is what the problem is, for sure. And that it totally makes building an API with this so much more complex than it should have to be. Ugh, express.

I'm trying to build more generic middleware, that doesn't need to know what path it's responding to; it will just examine `req.swagger.operation` and go from there. Is the path parameter only ever parsed by custom middleware, or does somewhere else in the swagger middleware parse it at least once (to validate it)? Maybe we can store it somewhere else on the request? That's kind of hacky, but needing to manually declare it is such a huge limitation.

That brings up another thought I had; a similar concept that swagger-tools has: being able to get the `value` on any param in the swagger object. So hitting `req.swagger.operation.parameters[x].value` would return the value, no need to examine `parameter.in` and then looking there on the request object (since it was already qualified by the swagger middleware).

**Comment by joshuajabbour** [comment]
Also, thanks for the thorough response!

**Comment by BigstickCarpet** [comment]
Thanks for the feedback.  I think you brought up a good point - that parsed path parameters should be available even to generic middleware that isn't tied to a specific path.   So, I just added that functionality.

Due to the way `req.params` works, there's no way for me to make it work with generic middleware.  So I added `req.pathParams` as well.  `req.params` still works for middleware that is bound to a parameterized path.  But `req.pathParams` will __always__ work, regardless of whether the middleware is bound to a path or not.  More details in [the docs](https://github.com/BigstickCarpet/swagger-express-middleware/blob/master/docs/middleware/parseRequest.md#phase-3---path-parsing).
